Rework seed alerts to consider all possible conditions

vicwicker opened the pull request:
**How to categorize this PR?**

/area monitoring
/kind enhancement

**What this PR does / why we need it**:

The list of current alerts for the seeds is incomplete. Furthermore, we detected some of the current alerts have misconfigurations that prevent them from working correctly. For example, some of them do not deal properly with flapping, or some others are set the `shoot` topology, which are not sent to the right receiver in the Alertmanager configuration.

This PR proposes grouping all seed conditions in one single alert. The benefits are twofold. First, the list of alerts is now complete. Second, having a single alert simplifies alert maintenance and reduces the risk of misconfigurations due to alert duplication.  

This PR also takes into account that seeds can be shoots as well and therefore queries both metrics `garden_seed_condition` and `garden_shoot_condition`. The seed conditions also show up in the shoot resource when the seed is a shoot but this does not pose a problem. On the contrary, by querying these two condition metrics, both managed and unmanaged seeds (e.g., soils) are tackled by the same alert.

The new alert is muted on the weekends for now. We first want to get an idea of how it behaves on canary and live. We might choose to silent it in the Alertmanager directly if it becomes too noisy but the ultimate goal is to unmute it in a follow-up PR.

Finally, please note we attempted to have individual alerts until commit `0a8a328`, when we iterated once again over the work in progress and decided to group all alerts. In consequence, this commit starts over with a clean slate. Nonetheless, we preserved previous commit history. 

**Special notes for your reviewer**:

/cc @istvanballok @rickardsjp 

**Release note**:
```other operator
Introduce a unified single alert for all seed conditions. Previous seed alerts `GardenletDown`, `GardenletUnknown`,  `SeedAPIServerUnavailable`, `SeedControlPlaneUnhealthy` and `SeedSystemComponentsUnhealthy` are removed.
```

vicwicker committed:
Revisit topology for current alerts

- Topology `shoot` to refer a seed is incorrect.
- Change topology from `garden` to `seed` for
  those alerts checking seed conditions.

vicwicker committed:
Remove unnecessary labels `is_compliant` and `has_user_errors`

We do not want to suppress alerts on seeds that are managed by the
Gardener team, even if the seeds are not compliant or the errors are
user errors.

vicwicker committed:
Make sure `max_over_time` is used in the alerts in case of flapping

Without it, flapping conditions would probably go undetected because
the alerts would be reset each time.

For example, when the seed is reconciling its artifacts, some of its
managed resources can enter the progressing state for a brief period,
and to avoid alerts from being restarted, the `max_over_time` function
is now introduced to smooth out this behaviour. This is combined with
the subquery syntax [1] to query over the last minutes only.

This change is also reducing the time before firing an alert in the `for`
entry from 30m to 10m for some alerts to alert more aggressively if any of
those conditions fail. An exception is the alert `SeedAPIServerUnavailable`,
which was already firing after 2m, but now changed to 3m to preserve the
same behaviour after the change introducing `max_over_time` to smooth
out possible flappings.

[1] https://prometheus.io/docs/prometheus/latest/querying/basics/#subquery

vicwicker committed:
Add new alerts to cover all possible seed conditions

vicwicker committed:
Homogenize summary and description for seed condition alerts

- Use common text for all alerts regarding seed conditions
- Fix description for the APIServerAvailable alert by adding
  the list of active alerts as in other alerts by `garden_shoot_condition`

vicwicker committed:
Make sure all seed alerts are not muted on the weekends

`mute_on_weekend` should be used for shoot alerts that are
customer issues and for which the Gardener team can't really
help. However, the Gardener team has full control of the
seeds so alerts on those should always be reacted upon: if
seed are having issues, then shoots can't be correctly managed.

vicwicker committed:
Fix description for the APIServerAvailable alert

Add the list of active alerts as in other alerts by `garden_shoot_condition`

vicwicker committed:
Add promtool tests for new seed condition alerts

The different condition states are mapped to numbers by the
`gardener-metrics-exporter`:

```
2: Progressing
1: True
0: False
-1: Unknown
```

vicwicker committed:
Add unit test for multiple shoots and alerts

This unit test shows that, if two different seeds, have failing
conditions then two different alerts pop up as well.

vicwicker committed:
Use `last_over_time` instead of `max_over_time`

The usage of a `<aggregate>_over_time` function is only required to
check if there is data within the queried time window that triggers
the alert. If the query return data, then the value is always true.
Otherwise, the query returns nothing.

Therefore, it is meaningless to calculate the maximum over a set
of "trues". Instead, using `last_over_time` is a better approach
to check if the query returned data.

vicwicker committed:
Mute new alert on seed conditions

Mute this alert on weekends as we gain experience on how noisy
it will be on live and canary. The goal, though, is that we
eventually unmute it.

vicwicker committed:
Alert after 10 minutes instead of 6 minutes

After testing this change on dev, we found alerting becomes
a bit too noisy so we choose to increase the for clause up
to 10 minutes again.

vicwicker committed:
Remove empty line at the end of the alert description

For better alert formatting, alert descriptions are not supposed to
contain an empty line at the very end.

istvanballok commented on pkg/component/observability/monitoring/prometheus/garden/testdata/seed.prometheusrule.test.yaml:
```diff
@@ -0,0 +1,31 @@
+    input_series:
+      - series: garden_seed_condition{condition = "ExtensionsReady",
+                                      name      = "seed-flapping"}
+        values: 0x2 2 0 -1 2 0x2 2 0
```
Maybe if the repetition is so low, we could also resolve the pattern for simplicity:


```
0x2 2 0 -1 2 0x2 2 0
```
-->
```
0 0 2 0 -1 2 0 0 2 0
```

Maybe for readability, we could mention in a comment here how the states are mapped to numbers by the `gardener-metrics-exporter`. https://github.com/gardener/gardener-metrics-exporter/blob/436fb84907a3778b0e90287155fe32b676942d2c/pkg/metrics/utils.go#L28-L39

> 2: Progressing
> 1: True
> 0: False
> -1: Unknown

vicwicker replied:
I addressed this by mentioning it in the first commit adding a promtool test: [08f99a4](https://github.com/gardener/gardener/pull/9750/commits/08f99a40a1c470c526299752584ca5fa51017ee7) so it helps when going through the whole commit history.

istvanballok replied:
As discussed off line, maybe add this information as a comment to the yaml test file as well, to improve the readability of the test for humans by resolving the "magic numbers" to concepts like Progressing/True/False/Unknown.

istvanballok commented on pkg/component/observability/monitoring/prometheus/garden/testdata/seed.prometheusrule.test.yaml:
```diff
@@ -2,20 +2,20 @@ rule_files:
 
 tests:
-  - name: SeedExtensionsNotReady:SmoothOutFlapping
+  - name: SeedConditionFailing:SmoothOutFlapping
```
Maybe split the commit `Use one single alert for all conditions instead` into 2 commits so that the diff introducing the new `SeedConditionFailing` alert is not "hidden" in the overall diff.

istvanballok reviewed (commented):
/lgtm with some minor comments for consideration

vicwicker committed:
[review] Address review comments

- Unroll the time series in the tests so it's simpler.
- Add comment on how conditions states are mapped into numbers.

vicwicker committed:
Use one single alert for all conditions: Remove previous alerts

This commit and the next discard most of previous work. The fundamental
change is that all seed condition are grouped into one alert only,
instead of having a specific alert for condition.

For the sake of change diff, this change is split in two commits: this
one removes previous alerts, and the next adds the new single alert.

vicwicker committed:
Use one single alert for all conditions: Add new alert

This and previous commit discard most of previous work. The fundamental
change is that all seed condition are now grouped into one alert only,
instead of having a specific alert for condition.

For the sake of change diff, this change is split in two commits: the
previous one removed previous alerts, and this adds the new single alert.

istvanballok reviewed (approved):
/lgtm
